src/main/java/com/yegor256/tojos/MnSynchronized.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2016-2022 Yegor Bugayenko
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.yegor256.tojos;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * This class is thread-safe.
+ *
+ * @since 0.3.0
+ */
+public final class MnSynchronized implements Mono {
+
+    /**
+     * The wrapped mono.
+     */
+    private final Mono wrapped;
+
+    /**
+     * Ctor.
+     *
+     * @param mono The mono
+     */
+    public MnSynchronized(final Mono mono) {
+        this.wrapped = mono;
+    }
+
+    @Override
+    public Collection<Map<String, String>> read() {
+        synchronized (this.wrapped) {
```

Comment:
You can use `ReadWriteLock` instead, it might increase an overall performance:
1. For read operations you need only `writeLock`
2. For write operations you need both `writeLock` and `readLock`

Reply:
@volodya-lombrozo have a look, please

Reply:
@volodya-lombrozo have a look, please

Reply:
@l3r8yJ It seems I've confused you. 

When you get `write` lock calling `writeLock().lock()` it doesn't allow reading too. From documentation:

> Acquires the write lock.
Acquires the write lock if neither the read nor write lock are held by another thread and returns immediately, setting the write lock hold count to one.
If the current thread already holds the write lock then the hold count is incremented by one and the method returns immediately.
If the lock is held by another thread then the current thread becomes disabled for thread scheduling purposes and lies dormant until the write lock has been acquired, at which time the write lock hold count is set to one.

As for `readLock().lock()` - it allows to read in parallel, but you can't write.
>Acquires the read lock.
Acquires the read lock if the write lock is not held by another thread and returns immediately.
If the write lock is held by another thread then the current thread becomes disabled for thread scheduling purposes and lies dormant until the read lock has been acquired.

Please, take a look on that [SO answer](https://stackoverflow.com/questions/46682896/how-to-add-locks-to-read-and-write-from-file-methods) and how people use ReadWriteLock. In other words, you have to use `writeLock().lock()`  in place where are you are using write operation and `readLock().lock()` where you are using read operation and `ReadWriteLock ` will behave as we want.

Reply:
@volodya-lombrozo yes, it all makes more sense now, thanks

Reply:
@volodya-lombrozo have a look, please
I've tried using the `synchronized` keyword lock with `wrapped` object, with `this`, with the `private final static Object LOCK = MnSynchronized.class` lock, with `ReadWriteLock` that you advised. Neither of these implementations passes the test, but if I use in line `64`:
```
synchronized (sync) {
    rows = sync.read();
    rows.add(row);
    sync.write(rows);
}
```
it always passes, even without the `MnSynchronized` decorator, do you know what could be the cause?